dot/core/service.go

```diff
@@ -321,59 +343,59 @@ func (s *Service) safeMsgSend(msg network.Message) error {
 func (s *Service) safeBabeKill() error {
 	s.lock.Lock()
 	defer s.lock.Unlock()
-	if s.closed {
+	// thread safe way to check closed status
+	if atomic.LoadUint32(&s.closed) == uint32(0) {
 		return ErrServiceStopped
 	}
 	close(s.babeKill)
 	return nil
 }
 
 func (s *Service) handleBabeSession() {
-	for {
-		// wait for BABE epoch to complete
-		<-s.epochDone
+	// wait for BABE epoch to complete
+	s.epochDone.Add(1)
+	s.epochDone.Wait()
```

Comment:
this seems excessively complicated since there was no issue with the channel, and there isn't really much point to using a WaitGroup for 1 process. can you show a case when receiving from a channel would cause a panic or other issue you are talking about? if not I would prefer it to be changed back

Reply:
Well, that was the whole point of the PR. Using channels is not indicated, WaitGroups behave way better. I strongly recommend we do not go back to that since this is working much better as we see on the improvements.